executor/insert_common.go

```diff
@@ -224,9 +236,121 @@ func insertRows(ctx context.Context, base insertCommon) (err error) {
 			}
 		}
 	}
+	// Fill the batch allocated autoIDs
+	rows, err = e.autoIDAllocN(ctx, rows)
+	if err != nil {
+		return err
+	}
 	return base.exec(ctx, rows)
 }
 
+func (e *InsertValues) autoIDAllocN(ctx context.Context, rows [][]types.Datum) ([][]types.Datum, error) {
+	retryInfo := e.ctx.GetSessionVars().RetryInfo
+	length := len(e.cache)
+	for i := 0; i < length; i++ {
+		switch e.cache[i].kind {
+		case AutoIDNull:
+			// Find consecutive num.
+			var cnt int
+			var start int
+			start = i
+			for i < length {
+				if e.cache[i].kind == AutoIDNull {
+					cnt++
+					i++
+				} else {
+					break
+				}
+			}
+			// Fix the index i
+			i--
+			// Alloc batch N consecutive autoIDs.
+			recordIDs, err := table.AllocBatchAutoIncrementValue(ctx, e.Table, e.ctx, cnt)
+			if e.filterErr(err) != nil {
+				return nil, err
+			}
+			// Distribute autoIDs to rows.
+			var d types.Datum
+			var c *table.Column
+			for j := 0; j < cnt; j++ {
+				offset := j + start
+				rowIdx := e.cache[offset].rowIdx
+				colIdx := e.cache[offset].colIdx
+
+				d = e.cache[offset].datum
+				c = e.Table.Cols()[colIdx]
+
+				// It's compatible with mysql. So it sets the first allocated autoID to the lastInsertId.
+				if e.lastInsertID == 0 {
+					e.lastInsertID = uint64(recordIDs[0])
+				}
+
+				d.SetAutoID(recordIDs[j], c.Flag)
+				retryInfo.AddAutoIncrementID(recordIDs[j])
+
+				// The value of d is adjusted by auto ID, so we need to cast it again.
+				d, err := table.CastValue(e.ctx, d, c.ToInfo())
+				if err != nil {
+					return nil, err
+				}
+				// Handle the bad null error.
+				if d, err = c.HandleBadNull(d, e.ctx.GetSessionVars().StmtCtx); err != nil {
+					return nil, err
+				}
+				rows[rowIdx][colIdx] = d
+			}
+		case AutoIDRebase:
+			// Rebase action
+			rowIdx := e.cache[i].rowIdx
+			colIdx := e.cache[i].colIdx
+
+			// recordID has been casted in evalRow
+			recordID := e.cache[i].recordID
+			d := e.cache[i].datum
+			c := e.Table.Cols()[colIdx]
+
+			err := e.Table.RebaseAutoID(e.ctx, recordID, true)
+			if err != nil {
+				return nil, err
+			}
+			e.ctx.GetSessionVars().StmtCtx.InsertID = uint64(recordID)
+			retryInfo.AddAutoIncrementID(recordID)
+
+			// Handle the bad null error.
+			if d, err = c.HandleBadNull(d, e.ctx.GetSessionVars().StmtCtx); err != nil {
+				return nil, err
+			}
+			// Cause d may changed in HandleBadNull, do assignment here
+			rows[rowIdx][colIdx] = d
+		case AutoIDZero:
+			// Don't change value 0 to auto id, if NoAutoValueOnZero SQL mode is set.
```

Comment:
```suggestion
			// Don't change value 0 to auto id, if `NO_AUTO_VALUE_ON_ZERO` SQL mode is set.
```
Is this a TODO?

Reply:
No, that means fill `0` to autoID column, just let it be.